saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/RelationsReconciler.kt

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2024 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.savedsites.impl
+
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.savedsites.store.SavedSitesEntitiesDao
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+interface RelationsReconciler {
+    fun reconcileRelations(
+        originalRelations: List<String>,
+        newFolderRelations: List<String>,
+    ): List<String>
+}
+
+@ContributesBinding(AppScope::class)
+class MissingEntitiesRelationReconciler @Inject constructor(
+    val savedSitesEntitiesDao: SavedSitesEntitiesDao,
+) : RelationsReconciler {
+    override fun reconcileRelations(
+        originalRelations: List<String>,
+        newFolderRelations: List<String>,
+    ): List<String> {
+        val missingEntities = (originalRelations - newFolderRelations.toSet()).filter {
+            savedSitesEntitiesDao.entityById(it) == null
```

Comment:
how big can this list be? accessing dao once per entity could take long. maybe you can store all entities in memory and compare them instead?